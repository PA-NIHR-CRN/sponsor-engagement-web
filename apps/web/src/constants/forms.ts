/**
 * Max amount of characters for a textarea
 */
export const TEXTAREA_MAX_CHARACTERS = 400

/**
 * Error messaging for different form submission failures
 */
export const FORM_ERRORS = {
  1: 'An unexpected error occurred whilst processing the form, please try again later.',
  2: 'The email address you have entered already exists for a user assigned to this organisation, so the user has not been added',
  3: "The user has been assigned to the organisation, but the invite email has not been sent as the user's email domain is not in the allowed list for testing",
  4: 'The entered email address is already a contact manager',
} as Record<number, string>

/**
 * Success messages for different form submissions
 */
export const FORM_SUCCESS_MESSAGES = {
  1: 'The study assessment was successfully saved',
  2: 'Your study data changes have been received. These will now be reviewed and applied to the study record. Until then, previous study data values will be displayed.', // Proposed study updates
  3: 'Your study data changes have been accepted.', // Direct study updates
} as Record<number, string>

export const UPDATE_FROM_SE_TEXT = 'Update from Sponsor Engagement Tool'
