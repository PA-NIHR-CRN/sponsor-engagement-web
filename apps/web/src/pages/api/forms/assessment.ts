import type { NextApiRequest } from 'next'
import { ZodError } from 'zod'
import type { Prisma } from 'database'
import type { AssessmentInputs } from '../../../utils/schemas'
import { assessmentSchema } from '../../../utils/schemas'
import { prismaClient } from '../../../lib/prisma'
import { withApiHandler } from '../../../utils/withApiHandler'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: AssessmentInputs
}

export default withApiHandler<ExtendedNextApiRequest>(async (req, res, session) => {
  try {
    if (req.method !== 'POST') {
      throw new Error('Wrong method')
    }

    const { studyId, status, furtherInformation, furtherInformationText } = assessmentSchema.parse(req.body)

    const furtherInformationInputs: Prisma.AssessmentFurtherInformationUncheckedCreateWithoutAssessmentInput[] = []

    // Create a furtherInformation record for each selected checkbox
    if (Array.isArray(furtherInformation)) {
      for (const id of furtherInformation) {
        furtherInformationInputs.push({
          furtherInformationId: Number(id),
        })
      }
    }

    // Create a record for the furtherInformation text field
    if (furtherInformationText) {
      furtherInformationInputs.push({
        furtherInformationText,
      })
    }

    const assessmentResult = await prismaClient.assessment.create({
      data: {
        createdById: session.user.id,
        studyId: Number(studyId),
        statusId: Number(status),
        furtherInformation: {
          createMany: {
            data: furtherInformationInputs,
          },
        },
      },
    })

    console.info(`Added assessment with id: ${assessmentResult.id}`)

    const studyResult = await prismaClient.study.update({
      where: {
        id: Number(studyId),
      },
      data: {
        isDueAssessment: false,
      },
    })

    console.info(`Updated study with id: ${studyResult.id}`)

    // Redirect back to study detail page
    if (String(req.query.returnUrl).includes(studyId)) {
      return res.redirect(302, `/studies/${studyId}?success=1`)
    }

    // Otherwise, redirect back to studies list page
    return res.redirect(302, `/studies?success=1`)
  } catch (error) {
    console.error(error)

    const studyId = req.body.studyId

    if (error instanceof ZodError) {
      // Create an object containing the Zod validation errors
      const fieldErrors: Record<string, string> = Object.fromEntries(
        error.errors.map(({ path: [fieldId], message }) => [`${fieldId}Error`, message])
      )

      // Insert the original values
      Object.keys(assessmentSchema.shape).forEach((field) => {
        if (req.body[field]) {
          fieldErrors[field] = req.body[field] as string
        }
      })

      delete fieldErrors.studyId

      const searchParams = new URLSearchParams({
        ...fieldErrors,
      })
      if (req.query.returnUrl) searchParams.append('returnUrl', String(req.query.returnUrl))

      return res.redirect(302, `/assessments/${studyId}/?${searchParams.toString()}`)
    }

    const searchParams = new URLSearchParams({
      fatal: '1',
    })
    if (req.query.returnUrl) searchParams.append('returnUrl', String(req.query.returnUrl))

    return res.redirect(302, `/assessments/${studyId}/?${searchParams.toString()}`)
  }
})
