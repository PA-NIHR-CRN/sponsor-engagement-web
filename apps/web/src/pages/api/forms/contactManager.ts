import crypto from 'node:crypto'

import { emailService } from '@nihr-ui/email'
import { logger } from '@nihr-ui/logger'
import { emailTemplates } from '@nihr-ui/templates/sponsor-engagement'
import type { NextApiRequest } from 'next'
import { ZodError } from 'zod'

import { Roles } from '@/constants'
import { MANAGE_CONTACT_MANAGERS_PAGE, SIGN_IN_PAGE } from '@/constants/routes'
import { prismaClient } from '@/lib/prisma'
import { getAbsoluteUrl } from '@/utils/email'
import type { ContactManagerAddInputs } from '@/utils/schemas'
import { contactManagerAddSchema } from '@/utils/schemas'
import { withApiHandler } from '@/utils/withApiHandler'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: ContactManagerAddInputs
}

export default withApiHandler<ExtendedNextApiRequest>([Roles.ContactManager], async (req, res, session) => {
  try {
    if (req.method !== 'POST') {
      throw new Error('Wrong method')
    }

    const { emailAddress } = contactManagerAddSchema.parse(req.body)

    // Get the contact manager sysref id from the database
    const { id: roleId } = await prismaClient.sysRefRole.findFirstOrThrow({
      where: {
        name: 'ContactManager',
        isDeleted: false,
      },
    })

    const {
      user: { id: requestedByUserId },
    } = session

    const registrationToken = crypto.randomBytes(24).toString('hex')

    let existingUser = await prismaClient.user.findUnique({
      include: { roles: true },
      where: {
        email: emailAddress,
      },
    })

    if (existingUser === null) {
      existingUser = await prismaClient.user.create({
        include: { roles: true },
        data: {
          email: emailAddress,
        },
      })
      // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison -- false positive for constant in conditional
    } else if (existingUser.roles.find((x) => x.roleId === Roles.ContactManager && x.isDeleted === false)) {
      const searchParams = new URLSearchParams({ fatal: '4' })
      return res.redirect(302, `${MANAGE_CONTACT_MANAGERS_PAGE}?${searchParams.toString()}`)
    }

    const shouldUpdateRegistrationToken = (existingUser.identityGatewayId ?? null) === null

    const isPreviousContactManager =
      // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison -- false positive for constant in conditional
      existingUser.roles.find((x) => x.roleId === Roles.ContactManager && x.isDeleted === true)
    if (isPreviousContactManager) {
      logger.info(`Re-adding contact manager role for ${existingUser.email}`)
    }

    const user = await prismaClient.user.update({
      where: {
        email: emailAddress,
      },
      data: {
        ...(shouldUpdateRegistrationToken && {
          registrationToken,
          registrationConfirmed: false,
        }),
        roles: {
          ...(!isPreviousContactManager && {
            create: {
              roleId,
              createdById: requestedByUserId,
              updatedById: requestedByUserId,
              isDeleted: false,
            },
          }),
          ...(isPreviousContactManager && {
            update: {
              where: {
                userId_roleId: {
                  roleId,
                  userId: existingUser.id,
                },
                isDeleted: true,
              },
              data: {
                updatedById: requestedByUserId,
                isDeleted: false,
              },
            },
          }),
        },
      },
    })

    const isNewUser = Boolean(user.registrationToken) && !user.registrationConfirmed

    const savedRegistrationToken = user.registrationToken

    await emailService.sendEmail({
      to: emailAddress,
      subject: 'NIHR RDN has invited you to be a Contact Manager for the Sponsor Engagement tool',
      htmlTemplate: emailTemplates['contact-manager-assigned.html.hbs'],
      textTemplate: emailTemplates['contact-manager-assigned.text.hbs'],
      templateData: {
        signInLink: getAbsoluteUrl(`${SIGN_IN_PAGE}${isNewUser ? `?registrationToken=${savedRegistrationToken}` : ``}`),
      },
    })

    return res.redirect(302, `${MANAGE_CONTACT_MANAGERS_PAGE}?success=1`)
  } catch (error) {
    logger.error(error)

    if (error instanceof ZodError) {
      // Create an object containing the Zod validation errors
      const fieldErrors: Record<string, string> = Object.fromEntries(
        error.errors.map(({ path: [fieldId], message }) => [`${fieldId}Error`, message])
      )

      // Insert the original values
      Object.keys(contactManagerAddSchema.shape).forEach((field) => {
        if (req.body[field]) {
          fieldErrors[field] = req.body[field] as string
        }
      })

      delete fieldErrors.emailAddress

      const searchParams = new URLSearchParams({ ...fieldErrors })

      return res.redirect(302, `${MANAGE_CONTACT_MANAGERS_PAGE}?${searchParams.toString()}`)
    }

    if ((error as Error).stack?.includes('MessageRejected')) {
      const searchParams = new URLSearchParams({ fatal: '3' })
      return res.redirect(302, `${MANAGE_CONTACT_MANAGERS_PAGE}?${searchParams.toString()}`)
    }

    const searchParams = new URLSearchParams({ fatal: '1' })

    return res.redirect(302, `${MANAGE_CONTACT_MANAGERS_PAGE}?${searchParams.toString()}`)
  }
})
