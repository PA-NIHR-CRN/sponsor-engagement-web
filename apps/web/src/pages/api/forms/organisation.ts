import crypto from 'node:crypto'
import type { NextApiRequest } from 'next'
import { ZodError } from 'zod'
import { logger } from '@nihr-ui/logger'
import { emailService } from '@nihr-ui/email'
import { emailTemplates } from '@nihr-ui/templates/sponsor-engagement'
import type { OrganisationAddInputs } from '../../../utils/schemas'
import { organisationAddSchema } from '../../../utils/schemas'
import { withApiHandler } from '../../../utils/withApiHandler'
import { getOrganisationById } from '../../../lib/organisations'
import { prismaClient } from '../../../lib/prisma'
import { Roles } from '../../../constants'
import {
  EXTERNAL_CRN_TERMS_CONDITIONS_URL,
  EXTERNAL_CRN_URL,
  SIGN_IN_PAGE,
  SUPPORT_PAGE,
} from '../../../constants/routes'
import { getAbsoluteUrl } from '../../../utils/email'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: OrganisationAddInputs
}

export default withApiHandler<ExtendedNextApiRequest>(Roles.ContactManager, async (req, res, session) => {
  try {
    if (req.method !== 'POST') {
      throw new Error('Wrong method')
    }

    const { emailAddress, organisationId } = organisationAddSchema.parse(req.body)

    const organisation = await getOrganisationById(Number(organisationId))

    if (!organisation) {
      throw new Error('No organisation found')
    }

    // Get the sponsor contact sysref id from the database
    const { id: roleId } = await prismaClient.sysRefRole.findFirstOrThrow({
      where: {
        name: 'SponsorContact',
        isDeleted: false,
      },
    })

    const {
      user: { id: contactManagerUserId },
    } = session

    const registrationToken = crypto.randomBytes(24).toString('hex')

    // Get existing user organisation record
    const userOrganisation = await prismaClient.userOrganisation.findFirst({
      where: {
        user: { email: emailAddress },
        organisation: { id: organisation.id },
      },
    })

    if (userOrganisation && !userOrganisation.isDeleted) {
      logger.error(
        'Tried to add contact with email %s to organisation %s, but active relationship already exists',
        emailAddress,
        organisation.name
      )
      const searchParams = new URLSearchParams({ fatal: '2' })
      return res.redirect(302, `/organisations/${organisationId}/?${searchParams.toString()}`)
    }

    if (userOrganisation) {
      logger.info('Re-adding contact with email %s to organisation %s', emailAddress, organisation.name)
    }

    // Add user to organisation
    const { name: organisationName, users } = await prismaClient.organisation.update({
      where: {
        id: Number(organisationId),
      },
      include: {
        users: {
          include: {
            user: true,
          },
        },
      },
      data: {
        users: {
          ...(userOrganisation && {
            update: {
              where: {
                id: userOrganisation.id,
              },
              data: {
                isDeleted: false,
                updatedBy: { connect: { id: contactManagerUserId } },
              },
            },
          }),
          ...(!userOrganisation && {
            create: {
              createdBy: { connect: { id: contactManagerUserId } },
              updatedBy: { connect: { id: contactManagerUserId } },
              user: {
                connectOrCreate: {
                  // If a user does not exist, create the user and set a registration token
                  create: {
                    email: emailAddress,
                    registrationToken,
                    registrationConfirmed: false,
                  },
                  where: {
                    email: emailAddress,
                  },
                },
              },
            },
          }),
        },
      },
    })

    await prismaClient.user.update({
      where: {
        email: emailAddress,
      },
      data: {
        roles: {
          // If a user is not assigned the sponsor contact role, assign it
          createMany: {
            data: {
              roleId,
              createdById: contactManagerUserId,
              updatedById: contactManagerUserId,
            },
          },
        },
      },
    })

    const isNewUser = users.some((user) => {
      const { email, registrationConfirmed, registrationToken: token } = user.user
      return email === emailAddress && Boolean(token) && registrationConfirmed === false
    })

    const savedRegistrationToken = users.find((user) => user.user.email === emailAddress)?.user.registrationToken

    await emailService.sendEmail({
      to: emailAddress,
      subject: 'Assess the progress of your studies',
      htmlTemplate: emailTemplates['contact-assigned.html.hbs'],
      textTemplate: emailTemplates['contact-assigned.text.hbs'],
      templateData: {
        organisationName,
        crnLink: EXTERNAL_CRN_URL,
        signInLink: getAbsoluteUrl(`${SIGN_IN_PAGE}${isNewUser ? `?registrationToken=${savedRegistrationToken}` : ``}`),
        requestSupportLink: getAbsoluteUrl(SUPPORT_PAGE),
        termsAndConditionsLink: EXTERNAL_CRN_TERMS_CONDITIONS_URL,
      },
    })

    return res.redirect(302, `/organisations/${organisationId}?success=1`)
  } catch (error) {
    logger.error(error)

    const organisationId = req.body.organisationId

    if (error instanceof ZodError) {
      // Create an object containing the Zod validation errors
      const fieldErrors: Record<string, string> = Object.fromEntries(
        error.errors.map(({ path: [fieldId], message }) => [`${fieldId}Error`, message])
      )

      // Insert the original values
      Object.keys(organisationAddSchema.shape).forEach((field) => {
        if (req.body[field]) {
          fieldErrors[field] = req.body[field] as string
        }
      })

      delete fieldErrors.organisationId

      const searchParams = new URLSearchParams({ ...fieldErrors })

      return res.redirect(302, `/organisations/${organisationId}/?${searchParams.toString()}`)
    }

    if ((error as Error).stack?.includes('MessageRejected')) {
      const searchParams = new URLSearchParams({ fatal: '3' })
      return res.redirect(302, `/organisations/${organisationId}/?${searchParams.toString()}`)
    }

    const searchParams = new URLSearchParams({ fatal: '1' })

    return res.redirect(302, `/organisations/${organisationId}/?${searchParams.toString()}`)
  }
})
