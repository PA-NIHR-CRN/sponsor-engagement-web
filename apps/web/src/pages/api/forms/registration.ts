import type { NextApiRequest, NextApiResponse } from 'next'
import { ZodError } from 'zod'
import { logger } from '@nihr-ui/logger'
import { authService } from '@nihr-ui/auth'
import type { RegistrationInputs } from '../../../utils/schemas'
import { registrationSchema } from '../../../utils/schemas'
import { prismaClient } from '../../../lib/prisma'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: RegistrationInputs
}

export default async (req: ExtendedNextApiRequest, res: NextApiResponse) => {
  try {
    if (req.method !== 'POST') {
      throw new Error('Wrong method')
    }

    const { password, registrationToken } = registrationSchema.parse(req.body)

    // Find contact information associated with the given registrationToken
    const user = await prismaClient.user.findFirst({
      where: {
        registrationToken,
        isDeleted: false,
      },
    })

    if (!user?.email) {
      throw new Error(`No email found associated with the registrationToken ${registrationToken}`)
    }

    // Create a new user in IDG
    const createUserResponse = await authService.createUser({
      password,
      emails: [user.email],
    })

    if (createUserResponse.success) {
      if (!createUserResponse.data.id) {
        throw new Error('Missing data from IDG createUser response')
      }

      const { id: identityGatewayId } = createUserResponse.data

      logger.info('Created user in IDG, updating user in local applicaton')

      const { id } = await prismaClient.user.update({
        where: {
          email: user.email,
          isDeleted: false,
        },
        data: {
          identityGatewayId,
          registrationConfirmed: true,
          registrationToken: null,
        },
      })

      logger.info(
        `Updated local user ${id} with identityGatewayId ${identityGatewayId}, now redirecting to confirmation page`
      )

      return res.redirect(302, `/register/confirmation`)
    }

    throw new Error('Failed to create user in IDG')
  } catch (error) {
    logger.error(error)

    const { registrationToken } = req.body

    if (error instanceof ZodError) {
      // Create an object containing the Zod validation errors
      const fieldErrors: Record<string, string> = Object.fromEntries(
        error.errors.map(({ path: [fieldId], message }) => [`${fieldId}Error`, message])
      )

      delete fieldErrors.registrationToken

      const searchParams = new URLSearchParams({ ...fieldErrors, registrationToken })
      return res.redirect(302, `/register?${searchParams.toString()}`)
    }

    const searchParams = new URLSearchParams({ fatal: '1', ...(registrationToken && { registrationToken }) })
    return res.redirect(302, `/register?${searchParams.toString()}`)
  }
}
