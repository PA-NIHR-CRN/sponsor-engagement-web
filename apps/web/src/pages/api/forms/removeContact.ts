import { emailService } from '@nihr-ui/email'
import { logger } from '@nihr-ui/logger'
import { emailTemplates } from '@nihr-ui/templates/sponsor-engagement'
import type { NextApiRequest } from 'next'

import { Roles } from '@/constants'
import { getUserOrganisationById } from '@/lib/organisations'
import { prismaClient } from '@/lib/prisma'
import type { OrganisationRemoveContactInputs } from '@/utils/schemas'
import { organisationRemoveContactSchema } from '@/utils/schemas'
import { withApiHandler } from '@/utils/withApiHandler'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: OrganisationRemoveContactInputs
}

export default withApiHandler<ExtendedNextApiRequest>(Roles.ContactManager, async (req, res, session) => {
  try {
    if (req.method !== 'POST') {
      throw new Error('Wrong method')
    }

    const { userOrganisationId } = organisationRemoveContactSchema.parse(req.body)

    const userOrganisation = await getUserOrganisationById(Number(userOrganisationId))

    if (!userOrganisation) {
      throw new Error('No user organisation found')
    }

    const {
      user: { id: contactManagerUserId },
    } = session

    // Remove user from organisation
    const { user, organisation } = await prismaClient.userOrganisation.update({
      where: {
        id: Number(userOrganisationId),
      },
      include: {
        organisation: true,
        user: true,
      },
      data: {
        isDeleted: true,
        updatedBy: { connect: { id: contactManagerUserId } },
      },
    })

    logger.info('Removed contact with email %s from organisation %s', user.email, organisation.name)

    if (user.email) {
      await emailService.sendEmail({
        to: user.email,
        subject: `NIHR RDN has removed you as a Sponsor contact for ${organisation.name}`,
        htmlTemplate: emailTemplates['contact-removed.html.hbs'],
        textTemplate: emailTemplates['contact-removed.text.hbs'],
        templateData: {
          organisationName: organisation.name,
        },
      })
    }

    return res.redirect(302, `/organisations/${userOrganisation.organisation.id}?success=2`)
  } catch (error) {
    logger.error(error)

    const userOrganisationId = req.body.userOrganisationId

    if ((error as Error).stack?.includes('MessageRejected')) {
      const searchParams = new URLSearchParams({ fatal: '3' })
      return res.redirect(302, `/organisations/remove-contact/${userOrganisationId}/?${searchParams.toString()}`)
    }

    const searchParams = new URLSearchParams({ fatal: '1' })

    return res.redirect(302, `/organisations/remove-contact/${userOrganisationId}/?${searchParams.toString()}`)
  }
})
