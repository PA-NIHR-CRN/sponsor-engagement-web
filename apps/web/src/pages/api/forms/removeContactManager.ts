import { logger } from '@nihr-ui/logger'
import type { NextApiRequest } from 'next'

import { Roles } from '@/constants'
import { CONTACT_MANAGERS_PAGE } from '@/constants/routes'
import { prismaClient } from '@/lib/prisma'
import type { RemoveContactManagerInputs } from '@/utils/schemas'
import { removeContactManagerSchema } from '@/utils/schemas'
import { withApiHandler } from '@/utils/withApiHandler'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: RemoveContactManagerInputs
}

export default withApiHandler<ExtendedNextApiRequest>([Roles.ContactManager], async (req, res, session) => {
  try {
    if (req.method !== 'POST') {
      throw new Error('Wrong method')
    }

    const { contactManagerUserId } = removeContactManagerSchema.parse(req.body)

    const {
      user: { id: requestedByUserId },
    } = session

    // Remove contact manager role from user.
    await prismaClient.userRole.update({
      where: {
        userId_roleId: {
          userId: Number(contactManagerUserId),
          roleId: Roles.ContactManager,
        },
        isDeleted: false,
      },
      data: {
        isDeleted: true,
        updatedById: requestedByUserId,
      },
    })

    logger.info(`user ${requestedByUserId} Removed contact manager role for User : ${contactManagerUserId}`)

    return res.redirect(302, `${CONTACT_MANAGERS_PAGE}?success=2`)
  } catch (error) {
    logger.error(error)

    const contactManagerUserId = req.body.contactManagerUserId

    if ((error as Error).stack?.includes('MessageRejected')) {
      const searchParams = new URLSearchParams({ fatal: '3' })
      return res.redirect(
        302,
        `/contact-managers/remove-contact-manager/${contactManagerUserId}/?${searchParams.toString()}`
      )
    }

    const searchParams = new URLSearchParams({ fatal: '1' })

    return res.redirect(
      302,
      `/contact-managers/remove-contact-manager/${contactManagerUserId}/?${searchParams.toString()}`
    )
  }
})
