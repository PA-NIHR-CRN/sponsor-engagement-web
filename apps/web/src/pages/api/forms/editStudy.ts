import type { Prisma } from 'database'
import type { NextApiRequest } from 'next'

import { Status, StudyUpdateRoute } from '@/@types/studies'
import { Roles, StudyUpdateType } from '@/constants'
import { UPDATE_FROM_SE_TEXT } from '@/constants/forms'
import { mapEditStudyInputToCPMSStudy, updateStudyInCPMS, validateStudyUpdate } from '@/lib/cpms/studies'
import { prismaClient } from '@/lib/prisma'
import { mapCPMSStatusToFormStatus } from '@/lib/studies'
import { constructDateObjFromParts } from '@/utils/date'
import type { EditStudyInputs } from '@/utils/schemas'
import { studySchema } from '@/utils/schemas'
import { withApiHandler } from '@/utils/withApiHandler'

export interface ExtendedNextApiRequest extends NextApiRequest {
  body: EditStudyInputs
}

export default withApiHandler<ExtendedNextApiRequest>(Roles.SponsorContact, async (req, res, session) => {
  const { ENABLE_DIRECT_STUDY_UPDATES } = process.env

  const enableDirectStudyUpdatesFeature = ENABLE_DIRECT_STUDY_UPDATES?.toLowerCase() === 'true'

  try {
    const studyData = studySchema.parse(req.body)
    const cpmsStudyInput = mapEditStudyInputToCPMSStudy(studyData)

    const { validationResult, error: validateStudyError } = await validateStudyUpdate(
      Number(studyData.cpmsId),
      cpmsStudyInput
    )

    // LSN for current state of study

    if (!validationResult) {
      throw new Error(validateStudyError)
    }

    // When feature flag is enabled, we use the response from the validate endpoint to determine the update type
    // Otherwise, we override this so we do not send any updates to CPMS
    const isDirectUpdate = enableDirectStudyUpdatesFeature
      ? validationResult.StudyUpdateRoute === StudyUpdateRoute.Direct
      : false

    const formattedPlannedOpeningDate = constructDateObjFromParts(studyData.plannedOpeningDate)
    const formattedActualOpeningDate = constructDateObjFromParts(studyData.actualOpeningDate)
    const formattedPlannedClosureDate = constructDateObjFromParts(studyData.plannedClosureDate)
    const formattedActualClosureDate = constructDateObjFromParts(studyData.actualClosureDate)
    const formattedEstimatedReopeningDate = constructDateObjFromParts(studyData.estimatedReopeningDate)

    const studyUpdate: Prisma.StudyUpdatesCreateInput = {
      studyStatus: isDirectUpdate ? studyData.status : null,
      studyStatusGroup: mapCPMSStatusToFormStatus(studyData.status),
      plannedOpeningDate: formattedPlannedOpeningDate,
      actualOpeningDate: formattedActualOpeningDate,
      plannedClosureToRecruitmentDate: formattedPlannedClosureDate,
      actualClosureToRecruitmentDate: formattedActualClosureDate,
      estimatedReopeningDate: formattedEstimatedReopeningDate,
      ukRecruitmentTarget: studyData.recruitmentTarget ? Number(studyData.recruitmentTarget) : undefined,
      comment: studyData.furtherInformation,
      study: {
        connect: {
          id: studyData.studyId,
        },
      },
      studyUpdateType: {
        connect: { id: isDirectUpdate ? StudyUpdateType.Direct : StudyUpdateType.Proposed },
      },
      createdBy: {
        connect: { id: session.user.id },
      },
      modifiedBy: {
        connect: { id: session.user.id },
      },
    }

    // Both proposed and direct changes are saved to SE
    await prismaClient.studyUpdates.create({
      data: studyUpdate,
    })

    if (isDirectUpdate) {
      // Only send additional note if new status is Suspended and not the original status
      // i.e. a status has been changed to Suspended
      const suspendedStatuses: string[] = [
        Status.SuspendedFromOpenToRecruitment,
        Status.SuspendedFromOpenWithRecruitment,
        Status.Suspended,
      ]
      const additionalNote =
        suspendedStatuses.includes(studyData.status) && !suspendedStatuses.includes(studyData.originalStatus ?? '')
          ? UPDATE_FROM_SE_TEXT
          : ''

      const { study, error: updateStudyError } = await updateStudyInCPMS(Number(studyData.cpmsId), {
        ...cpmsStudyInput,
        notes: additionalNote,
      })

      // LSN returned here and save to study updates
      if (!study) {
        throw new Error(updateStudyError)
      }
    }

    return res.redirect(302, `/studies/${studyData.studyId}?success=${isDirectUpdate ? 3 : 2}`)
  } catch (e) {
    const searchParams = new URLSearchParams({ fatal: '1' })
    const studyId = req.body.studyId

    return res.redirect(302, `/studies/${studyId}/edit?${searchParams.toString()}`)
  }
})

// Task 1: EditHistory - proposed GET CPMS response to include:
// [{
//   LSN: '3534'
//   modifiedDate: '2024-10-07 12:44:25.423' <- if direct change, no link to SE so don't know what date.
//   changes: [{
//     columnChanged: 'plannedOpeningDate' // Have this as CPMS field - mapping in UI needed for label
//     beforeValue: '2024-10-07 12:44:25.423'
//     afterValue: '2024-12-10 12:44:25.423'
//     id: '12212', // randomnly generated
//   }]
// }]

// Task 2: Saving LSN to study updates table
// Open q: Do study update after CPMS call

// Task 3: Save diff changes to existing or create new table
// 1. Edit study - fetch study from SE, compare to create diff object
// 2. Store - in existing or new table

// Task 4: Displaying
// 1. GET CPMS - for each edit history, look up in SE. If LSN (direct), use email and date. If no LSN (direct), use RDN and date from GET res
// 2. GET SE - for proposed change, already has email and date. Either fetch from existing or new table using studyUpdateId
// Map the above into a common type to be used by EditHistory component
// Direct + In direct changes will have 'Changed...'
// Proposed changes will have 'Proposed change...
