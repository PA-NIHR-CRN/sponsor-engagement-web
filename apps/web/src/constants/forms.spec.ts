import { FORM_ERRORS } from './forms'

test('Form errors', () => {
  expect(FORM_ERRORS).toMatchSnapshot()
})
