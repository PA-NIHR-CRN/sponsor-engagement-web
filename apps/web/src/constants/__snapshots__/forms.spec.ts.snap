// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form errors 1`] = `
{
  "1": "An unexpected error occurred whilst processing the form, please try again later.",
  "2": "The email address you have entered already exists for a user assigned to this organisation, so the user has not been added",
  "3": "The user has been assigned to the organisation, but the invite email has not been sent as the user's email domain is not in the allowed list for testing",
  "4": "The entered email address is already a contact manager",
}
`;
